### PR TITLE
Change boolean columns from char(1) to tinyint(1)

### LINKED ISSUE #4184: Use boolean fields for checkboxes
Since Contao 4.9.11 we fully support "boolean" database fields in the data container. In regards to https://github.com/contao/contao/pull/4179 it would thus be more convenient to use

```php
['type' => 'boolean', …]
```

instead of

```php
"char(1) COLLATE ascii_bin NOT NULL default ''"
```

in order to reduce the size of the field from 5 bytes to 1. `char` and `varchar` fields with `utf8mb4` collation currently require 5 bytes in total under InnoDB (4 + 1 for saving the length of the content).

I propose we change this in Contao 5.0. Note that this will need a migration as a database field cannot be changed from `char` (with empty content) to `tinyint` directly (since an empty string is not valid content for `tinyint` and thus will lead to an error in strict mode).

### PR DESCRIPTION
**Pull request by fritzmg**
Implements #4184

_Note:_ the migration automatically converts the empty string value of _all_ `char(1)` fields to `0`, if the DCA SQL definition is of type `boolean`. So this would also automatically be done for any extensions, if applicable.

### CODE REVIEW

File: core-bundle/src/Controller/FrontendModule/TwoFactorController.php
Comment (ausi) [inline core-bundle/src/Controller/FrontendModule/TwoFactorController.php:153]: If we read the user from the database would the value not still be fetched as `'1'` (or `'0'` if false)?

**Comment by fritzmg** [comment]
PHP stan complains about

```
Property Contao\User::$useTwoFactor (bool|string) does not accept int.
```

How should we handle this?

1. Change doc to `string|integer` and always use `0`/`1` in assignments.
2. Change all assignments from `0`/`1` to `false`/`true`.
3. Change doc to `string|integer|boolean`.
4. Change doc to `integer|boolean`.

Thoughts:

- When coming from the database, the value will either be a `string` or an `integer` under PDO, depending on the PHP version. However, since Contao 5 already requires PHP 8.1, we can assume that these values are always loaded as integers from the database, correct?
- When using the model, users will likely use `false`/`true` (in the future) when assigning these variables.

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (ausi) [inline core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]:
```suggestion
                if ('boolean' !== ($fieldConfig['sql']['type'] ?? null)) {
```
is_array check is not needed as this is handled by `??` too.

File: core-bundle/src/Controller/FrontendModule/TwoFactorController.php
Comment (fritzmg) [reply core-bundle/src/Controller/FrontendModule/TwoFactorController.php:153]: No, it will be `0` or `1` (with PDO under PHP 8.1).

**Comment by ausi** [comment]
> ~~Change doc to `string|integer` and always uses `0`/`1` in assignments.~~

~~I’d favor this one for now I think.~~

Changed my mind to:

> Change doc to `string|integer|boolean`.

**Comment by ausi** [comment]
> we can assume that these values are always loaded as integers from the database, correct?

Probably not because of mysqli :(

**Comment by fritzmg** [comment]
> > Change doc to `string|integer` and always uses `0`/`1` in assignments.
> 
> I’d favor this one for now I think.

It's not very intuitive though to have to write `$model->someBoolean = 1` instead of `true`.

**Comment by fritzmg** [comment]
I think we would need to go with option no. 3 then in that case? i.e. change the doc to `string|integer|boolean` because:

* when loading from the database the value can either be an integer (PDO) or a string (MySQLi)
* when setting values, users will likely want to use `true`/`false`

**Comment by ausi** [comment]
I think we should merge this one with `string|integer|boolean` as this can be changed to `boolean` in #4800 then.

**Review by ausi — APPROVED** [review]
❤️

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (ausi) [inline core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]: This needs to be changed, see #4803

**Review by ausi — CHANGES_REQUESTED** [review]

File: core-bundle/tests/EventListener/DataContainer/PreviewLinkListenerTest.php
Comment (leofeyer) [inline core-bundle/tests/EventListener/DataContainer/PreviewLinkListenerTest.php]: `$showUnpublished` is a boolean value already, isn‘t it?

File: core-bundle/tests/Routing/ResponseContext/CoreResponseContextFactoryTest.php
Comment (leofeyer) [inline core-bundle/tests/Routing/ResponseContext/CoreResponseContextFactoryTest.php]: Shouldn‘t this be `true` instead?

Comment (fritzmg) [reply core-bundle/tests/Routing/ResponseContext/CoreResponseContextFactoryTest.php]: My thinking was that we should simulate in the tests how the value would be in a real application. Under PHP 8.1 with PDO, the value will be `1`, not `true`. Except when #4800 gets merged, then it would be `true`.

**Comment by leofeyer** [comment]
As discussed, we want to wait for #4800 before we merge this PR.

**Comment by fritzmg** [comment]
Unfortunately I have hit a roadblock here. Doctrine seems to convert parameter values that are `false` to an empty string, when using

```php
$objDatabase->prepare("INSERT INTO tl_search %s")
            ->set($arrSet)
            ->execute();
```

With this PR the following error will occur when trying to log into the back end for example:

```
Doctrine\DBAL\Exception\DriverException:
An exception occurred while executing a query: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect integer value: '' for column `c50`.`tl_user`.`disable` at row 1

  at vendor\doctrine\dbal\src\Driver\API\MySQL\ExceptionConverter.php:119
  at Doctrine\DBAL\Driver\API\MySQL\ExceptionConverter->convert(object(Exception), object(Query))
     (vendor\doctrine\dbal\src\Connection.php:1814)
  at Doctrine\DBAL\Connection->handleDriverException(object(Exception), object(Query))
     (vendor\doctrine\dbal\src\Connection.php:1749)
  at Doctrine\DBAL\Connection->convertExceptionDuringQuery(object(Exception), '…', array(…), array())
     (vendor\doctrine\dbal\src\Connection.php:1055)
  at Doctrine\DBAL\Connection->executeQuery('…', array(…), array())
     (vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\Database\Statement.php:269)
  at Contao\Database\Statement->query('', array(…))
     (vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\Database\Statement.php:228)
  at Contao\Database\Statement->execute(1)
     (vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\User.php:297)
```

The functional tests will also fail [here](https://github.com/contao/contao/blob/eb37857c26de5fd3a230d583681e78ced0178ce4/core-bundle/src/Resources/contao/library/Contao/Search.php#L223-L225) for the same reason.

**Comment by fritzmg** [comment]
As briefly discussed with @ausi I am now casting any booleans to integer in `Statement::query`, so that Doctrine does not convert `false` to an empty string. However, if you use the Doctrine database connection directly, you will need to make of that sure yourself.

File: core-bundle/src/Command/UserPasswordCommand.php
Comment (ausi) [inline core-bundle/src/Command/UserPasswordCommand.php]:
```suggestion
                'pwChange' => $input->getOption('require-change'),
            ],
            ['username' => $input->getArgument('username')],
            [ParameterType::STRING, ParameterType::BOOLEAN, ParameterType::INTEGER, ParameterType::BOOLEAN]
```

something like that?

**Comment by fritzmg** [comment]
Btw.: https://bugs.php.net/bug.php?id=57157 😑 

> From the documentation of PDOStatement::execute()[1]:
> 
> > An array of values with as many elements as there are bound
> > parameters in the SQL statement being executed. All values are
> > treated as PDO::PARAM_STR.
> 
> Coercing TRUE to string yields "1"; coercing FALSE to string
> yields "". Thus, the reported behavior is expected.  Explicitly
> bind as PDO::PARAM_INT instead.
> 
> [1] <https://www.php.net/manual/en/pdostatement.execute.php>

**Review by ausi — DISMISSED** [review]

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (ausi) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]:
```suggestion
            Controller::loadDataContainer($tableName);
```

I think we should just remove the try catch block here, and fix this together with #4803 in another PR.

**Comment by fritzmg** [comment]
I still need to check if there are more places where we need to cast `bool` to `int` (or pass the parameter type).

**Comment by fritzmg** [comment]
_Note:_ I only checked a few more places. I'd vote to merge this as is and then fix any issue later on as discovered.

File: comments-bundle/src/Resources/contao/models/CommentsNotifyModel.php
Comment (aschempp) [inline comments-bundle/src/Resources/contao/models/CommentsNotifyModel.php:26]: are these really booleans now? Not integers?

File: core-bundle/src/Resources/contao/dca/tl_user.php
Comment (aschempp) [inline core-bundle/src/Resources/contao/dca/tl_user.php:893]: since this is a `save_callback`, it should probably set an integer value?

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: not sure I like this. We should include the type information of columns, then Doctrine would automatically convert the values according to the database abstraction (not all platforms use integer as boolean). Not sure if we can do that in all places though, maybe we only cast if the column is not given in `$arrTypes`?

File: comments-bundle/src/Resources/contao/models/CommentsNotifyModel.php
Comment (fritzmg) [reply comments-bundle/src/Resources/contao/models/CommentsNotifyModel.php:26]: Yes, see https://github.com/contao/contao/pull/4800

File: core-bundle/src/Resources/contao/dca/tl_user.php
Comment (fritzmg) [reply core-bundle/src/Resources/contao/dca/tl_user.php:893]: It will automatically be converted to `int` by Contao anyway.

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php
Comment (fritzmg) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: > Not sure if we can do that in all places though

We can in fact not do it with `Contao\Database\Statement::set()` currently. Also it would be cumbersome to do that in any case.

However yes, we could at least check for `empty($arrTypes)`.

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php
Comment (aschempp) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php]: now we can 😎 https://github.com/contao/contao/pull/4839
Yes it might be cumbersome, but that's just how Doctrine works. If you use boolean fields in DBAL, you must include type information for it to work. Maybe our model abstraction could automatically include it. Also, I'm fine with still casting if there is no type 🙃. Just be aware that params and types are NOT keyed, they are by order.

File: core-bundle/src/Command/UserListCommand.php
Comment (leofeyer) [inline core-bundle/src/Command/UserListCommand.php]: Shouldn‘t this be `true` instead? I know that the DB column expects `0` or `1` but this seems bad DX to me:

```
$user = $userModel->findBy('admin', 1);

dump($user->admin); // true
```

And if we want to change the value:

```
$user->admin = false;
$user->save();
```

File: core-bundle/src/Resources/contao/classes/Backend.php
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/Backend.php:700]: Shouldn‘t this be `true` here as well?

File: core-bundle/src/Resources/contao/drivers/DC_Table.php
Comment (leofeyer) [inline core-bundle/src/Resources/contao/drivers/DC_Table.php]:
```suggestion
		$this->save(!$this->varValue);
```

File: core-bundle/tests/Contao/ControllerTest.php
Comment (leofeyer) [inline core-bundle/tests/Contao/ControllerTest.php]: Why is this `0` here but `false` on line 157?

File: core-bundle/tests/Controller/SitemapControllerTest.php
Comment (leofeyer) [inline core-bundle/tests/Controller/SitemapControllerTest.php]: This should be `true`, shouldn‘t it? Because we are mocking a model here. (Seems like you have been using boolean values as of line 435 already?)

File: core-bundle/tests/EventListener/Security/TwoFactorFrontendListenerTest.php
Comment (leofeyer) [inline core-bundle/tests/EventListener/Security/TwoFactorFrontendListenerTest.php]: Why an integer here but booleans on line 264 and 268?

**Review by leofeyer — CHANGES_REQUESTED** [review]
This PR still seems very inconsistent to me. If I understand correctly, we have to use integers when working with DB queries and result objects, and booleans when working with models. Any exception to this rule, e.g. in a unit test, should be explicitly noted.

**Comment by fritzmg** [comment]
Yes, the inconsistency comes from the change in #4800 and I found it to not be as easy anymore to find all `0` and `1` that need to be `false` and `true` now.

File: core-bundle/src/Resources/contao/classes/Backend.php
Comment (fritzmg) [reply core-bundle/src/Resources/contao/classes/Backend.php:700]: The `$row` is a `\Contao\Database\Result` row and not the row of a Model.

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (rabauss) [inline core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]:
```suggestion
                if (\Doctrine\DBAL\Types\Types::BOOLEAN !== ($fieldConfig['sql']['type'] ?? null)) {
```
why not using the constant? could also be used in the dca, couldn't it?

Comment (fritzmg) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]: For 5.x this is ok. But if you support older Doctrine versions you can't use them as they had been moved once (just as a side note).

Comment (fritzmg) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]: I would leave the DCAs as they are for now and may be later switch to the constant in a separate PR.

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:258]: Missing `use ($arrTypes)` I think?

**Review by ausi — DISMISSED** [review]

File: comments-bundle/src/Resources/contao/models/CommentsModel.php
Comment (m-vo) [inline comments-bundle/src/Resources/contao/models/CommentsModel.php:122]: Should we also go for `=TRUE` and `=FALSE` in comparisons?
These are synonyms for `1` and `0` in MySQL: https://dev.mysql.com/doc/refman/8.0/en/boolean-literals.html

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (m-vo) [inline core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]:
```suggestion
            $this->connection->update($table, [$column => '0'], [$column => '']);
```

File: core-bundle/tests/Security/User/UserCheckerTest.php
Comment (m-vo) [inline core-bundle/tests/Security/User/UserCheckerTest.php:112]: The `$user->$start` property is supposed to be `string|integer`, but why needs this test to be changed?

**Review by m-vo — DISMISSED** [review]
LGTM :heart: 

I just left some minor questions/nitpicks.

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (fritzmg) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]: Why a string?

**Comment by fritzmg** [comment]
Note: the changes from https://github.com/contao/contao/pull/4881 also need to be integrated into the migration of this PR.

File: core-bundle/tests/Security/User/UserCheckerTest.php
Comment (fritzmg) [reply core-bundle/tests/Security/User/UserCheckerTest.php:112]: `$user->start` is a `varchar` in the database - and it will stay a string when loaded from the database. As this test simulates a model returned from the database, the value should be a string, not an integer.

File: comments-bundle/src/Resources/contao/models/CommentsModel.php
Comment (fritzmg) [reply comments-bundle/src/Resources/contao/models/CommentsModel.php:122]: I personally prefer `0`/`1` as it's shorter, but I am fine either way. /cc @contao/developers

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (m-vo) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]: Isn't the idea of the migration to prepare the database before the column types get changed? They will still be `CHAR(1)` at this point.

Comment (fritzmg) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]: I suppose so, not that it matters much in the end :)

Comment (m-vo) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php]: Yes, it was more of a mental help to help you understand when this is executed. But feel free to resolve. :slightly_smiling_face:

File: comments-bundle/src/Resources/contao/models/CommentsModel.php
Comment (leofeyer) [reply comments-bundle/src/Resources/contao/models/CommentsModel.php:122]: I also prefer the integers. 👍

**Review by ausi — DISMISSED** [review]

File: core-bundle/src/Resources/contao/dca/tl_page.php
Comment (leofeyer) [inline core-bundle/src/Resources/contao/dca/tl_page.php]:
```suggestion
			'options'                 => array('http://', 'https://'),
```

Adding the array keys is redundant, isn‘t it?

Comment (fritzmg) [reply core-bundle/src/Resources/contao/dca/tl_page.php]: I'd still prefer it in this case, but I'm fine either way.

Comment (fritzmg) [reply core-bundle/src/Resources/contao/dca/tl_page.php]: Ah - actually neither will work. Now that the values are normal integers, the widget will treat the labels as values. I think we'll need a load and save callback here. Or use `reference`.

Comment (m-vo) [reply core-bundle/src/Resources/contao/dca/tl_page.php]: Would `'isAssociative' => true` solve this?

Comment (fritzmg) [reply core-bundle/src/Resources/contao/dca/tl_page.php]: @m-vo I'll try that. Otherwise I fixed it in bfe3b49aca5f42bd38e6246dfb9fe6c1860efb06

Comment (fritzmg) [reply core-bundle/src/Resources/contao/dca/tl_page.php]: Yes, that works too 👍

**Comment by fritzmg** [comment]
~~There is still an issue with the `checkbox` widget. It currently gets `checked` even though the value is `0`.~~ // false alarm

**Review by m-vo — DISMISSED** [review]

File: core-bundle/src/Resources/contao/dca/tl_page.php
Comment (leofeyer) [inline core-bundle/src/Resources/contao/dca/tl_page.php:548]: Why the flag?

**Review by leofeyer — CHANGES_REQUESTED** [review]
There are still plenty of wrong values in the code base, I‘m afraid. A first quick search revealed the following:

- CalendarEventsModel.php line 269
- Dbafs.php line 516
- FormFieldModel.php line 201
- FormModel.php line 121
- ImageSizeItemModel.php line 95
- MemberGroupModel.php line 83 and 101
- MemberModel.php line 197
- Module.php line 476
- NewsModel.php line 238

We have to make sure to search for `='1'` and `=''` as well as `= '1'` and `= ''` to find all occurrences.

File: core-bundle/src/Resources/contao/dca/tl_page.php
Comment (fritzmg) [reply core-bundle/src/Resources/contao/dca/tl_page.php:548]: See https://github.com/contao/contao/pull/4797#discussion_r909607112 - otherwise the select won't work.

**Comment by fritzmg** [comment]
> We have to make sure to search for `='1'` and `=''` as well as `= '1'` and `= ''` to find all occurrences.

Also `=> ''` and `=> '1'`.

I made another sweep.

File: core-bundle/src/Migration/Version500/BooleanFieldsMigration.php
Comment (rabauss) [inline core-bundle/src/Migration/Version500/BooleanFieldsMigration.php:99]: Are fields of extensions included too, right?

Comment (fritzmg) [reply core-bundle/src/Migration/Version500/BooleanFieldsMigration.php:99]: Yes, see my initial comment. However in the new version, any fields of _dynamcially_ added DCAs (like Catalog Manager or MetaModels) would not be included.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**
